### PR TITLE
Projects/piloting feature/keypoints integration piloting

### PR DESCRIPTION
**Pull request by grizzi**
This should be merged after #79 

It makes sure we can simulated keypoints detection in simulation. 
The state machine has an additional observation state which finds the best point of view, such that given a known size of the object and its approximate pose (from the tf tree) the object is in the field of view. 

Tested in simulation. Make sure to pull `smb_common` from my fork and build `smb_gazebo`. The changes there concern the odometry plugin. This is publishing the tf from a fixed robot link to gazebo world. This way we have the "gazebo aware" world link of which we talk in PR #79. This is also the reason why we do not have to publish ourself the dummy transform `world` to `tracking_camera_odom`

### CODE REVIEW

File: moma_mission/config/state_machine/piloting.yaml [hunk 1]
```diff
@@ -45,21 +45,29 @@ HOME_ROBOT_FINAL:
...
   default_outcome: None
-  object_position_estimate: [0.8, 0.0, 0.0]
-  phi_range: [0, 0.4]
+  sample_name: observation
+  map_frame: world
+  object_frame: valve_gt
+  camera_frame: hand_eye_color_optical_frame
+  controlled_frame: tool_frame
+  object_size: 0.20
+  camera_info_topic: "/hand_eye/color/camera_info"
```
Comment (jk-ethz) [inline moma_mission/config/state_machine/piloting.yaml:54]: This should probably be `/camera_info` which is our own "corrected" calibration.

File: moma_mission/src/moma_mission/states/observation.py [hunk 2]
```diff
@@ -66,3 +67,138 @@ def run(self):
...
+
+            sample_pose_array = PoseArray()
+            for theta, phi in zip(thetas, phis):
+                coords = self._spherical_to_cartesian(d, theta, phi)
+
+                z_axis_dir = -tf.transformations.unit_vector(coords)
+                x_axis_dir_desired = self.heading
+                y_axis_dir = tf.transformations.unit_vector(np.cross(z_axis_dir, x_axis_dir_desired))
+                x_axis_dir = np.cross(y_axis_dir, z_axis_dir)
+                assert np.allclose(x_axis_dir, tf.transformations.unit_vector(x_axis_dir))
```
Comment (jk-ethz) [inline moma_mission/src/moma_mission/states/observation.py:155]: Do you think it is feasible to inherit from `SphericalSamplerState` (maybe outsource some functionality to separate functions there that can be reused) and avoid some duplication?

File: moma_mission/src/moma_mission/states/observation.py [hunk 3]
```diff
@@ -66,3 +67,138 @@ def run(self):
...
+
+        # how big in the image the object should look like. 1.0 means getting the closest to the object
+        self.image_fill_ratio = self.get_scoped_param("image_fill_ratio")
+
+        # needed for FOV computation
+        self.camera_info_topic = self.get_scoped_param("camera_info_topic", "/camera_info")
+        self.heading = self.get_scoped_param("heading", [1, 0, 0])
+        
+        # how much to look lateraly across all candiate observation poses (it is the phi angle in sperical coords)
+        self.lateral_view_angle = np.deg2rad(self.get_scoped_param("lateral_view_angle", 20))
```
Comment (jk-ethz) [inline moma_mission/src/moma_mission/states/observation.py:109]: I think we should have a common definition of how we specify angle parameters throughout the MoMa stack. I have seen a mix of both, degrees and radians. For the spherical observer I defaulted to radians.

File: moma_mission/src/moma_mission/states/transform_visitor.py [hunk 4]
```diff
@@ -13,7 +13,7 @@ class TransformVisitorState(StateRosControl):
     def __init__(self, ns):
         StateRosControl.__init__(self, ns=ns)
 
-        self.world_frame = self.get_scoped_param("world_frame", "world")
+        self.control_frame = self.get_scoped_param("control_frame", "odom")
```
Comment (jk-ethz) [inline moma_mission/src/moma_mission/states/transform_visitor.py:16]: I don't quite understand yet why in MPC it was called `reference_frame` and here `control_frame`.

File: moma_mission/src/moma_mission/states/observation.py [hunk 5]
```diff
@@ -66,3 +67,138 @@ def run(self):
...
+            rospy.sleep(1.0)
+        
+        if self.attempts < self.num_samples:
+            self.pose_broadcaster.sendTransform(self.sample_poses[self.attempts])
+            self.attempts += 1
+            rospy.sleep(2.0)
+            return 'Completed'
+        else:
+            rospy.logwarn("No more point of views.")
+            return 'Failure'
```
Comment (jk-ethz) [inline moma_mission/src/moma_mission/states/observation.py:204]: New line at EOF

File: moma_mission/src/moma_mission/states/observation.py [hunk 6]
```diff
@@ -66,3 +67,138 @@ def run(self):
...
+        if self.first_run:
+            w = np.min([self.camera_info.height, self.camera_info.width])
+            f = np.max([self.camera_info.K[0], self.camera_info.K[4]])
+            d = self.object_size * f / ( w * self.image_fill_ratio)
+
+            T_map_obj = self.get_transform(self.map_frame, self.object_frame)
+            T_cam_ctrl = self.get_transform(self.camera_frame, self.controlled_frame)
+        
+            thetas = [0.0] + [i * np.pi/4 for i  in range(8)]
+            phis = [0.0] + [self.lateral_view_angle] * 8
```
Comment (jk-ethz) [inline moma_mission/src/moma_mission/states/observation.py:145]: 8 -> `self.num_samples + 1`

File: moma_mission/src/moma_mission/states/observation.py [hunk 7]
```diff
@@ -66,3 +67,138 @@ def run(self):
...
+
+        if self.first_run:
+            w = np.min([self.camera_info.height, self.camera_info.width])
+            f = np.max([self.camera_info.K[0], self.camera_info.K[4]])
+            d = self.object_size * f / ( w * self.image_fill_ratio)
+
+            T_map_obj = self.get_transform(self.map_frame, self.object_frame)
+            T_cam_ctrl = self.get_transform(self.camera_frame, self.controlled_frame)
+        
+            thetas = [0.0] + [i * np.pi/4 for i  in range(8)]
```
Comment (jk-ethz) [inline moma_mission/src/moma_mission/states/observation.py:144]: 8 -> `self.num_samples + 1`

File: moma_mission/src/moma_mission/states/observation.py [hunk 8]
```diff
@@ -66,3 +67,138 @@ def run(self):
 
         rospy.sleep(2.0)
         return 'Completed'
+
+class FOVSamplerState(StateRos):
```
Comment (jk-ethz) [inline moma_mission/src/moma_mission/states/observation.py:71]: Nice additional option for state sampling!

**Comment by jk-ethz** [comment]
Before merge, need to align urdf to custom publisher of extrinsics (as per real robot and `easy_handeye` publish_calibration.launch) using configurable extrinsics in launch. Test that this does not break the real extrinsics setup.

File: moma_mission/src/moma_mission/states/observation.py (hunk 3 above)
Comment (grizzi) [reply moma_mission/src/moma_mission/states/observation.py:109]: For the state machine params I would rather use rad and use the `_deg` suffix. While in mpc those params are static, is very likely that here we want to repeadatily try values, and degrees are more difficult to mess-up and we have a more realistic intuition of what they mean.

File: moma_demos/piloting_demo/launch/perception.launch [hunk 9]
```diff
@@ -1,33 +1,38 @@
...
     <arg name="load_bag_filepath" default="false"/>
-    <arg name="calibration_file"  default="$(env HOME)/calibration/intrinsics.yaml"/>
+    <arg name="calibration_file"  default="$(arg calibration_file)"/>
     <arg name="image_topic"       default="$(arg image_topic)"/>
     <arg name="depth_topic"       default="$(arg depth_topic)"/>
     <arg name="calibration_topic" default="$(arg calibration_topic)"/>
     <arg name="base_frame"        default="$(arg base_frame)"/>
     <arg name="camera_frame"      default="$(arg camera_frame)"/>
-    <arg name="model_name"        default="valve_depth"/>
+    <arg name="model_name"        default="valve"/>
```
Comment (jk-ethz) [inline moma_demos/piloting_demo/launch/perception.launch:27]: @grizzi Note that I updated `object_keypoints_ros`.
I think it makes more sense to have a general "valve" name and in the end also rename the "valve_depth.pt" to "valve.pt". This way we have a unified naming, also for "valve.json" for the keypoints, etc.

**Pull request merged**